scipy/sparse/sparsetools/csr.h

```diff
@@ -254,32 +257,60 @@ void csr_tobsr(const I n_row,
 }
 
 
+template <class I>
+static inline void check_bounds(const I index,
+                                const I size)
+{
+    if NPY_UNLIKELY(index >= size) {
+        std::ostringstream err;
+        err << "internal error: index out of bounds: " << index
+            << ", size: " << size;
+        throw std::runtime_error(err.str());
```

Comment:
I think we'd like to convert this to a Python exception - can be done once we've decided this is the right approach.

It'd be something like https://stackoverflow.com/questions/32391561/how-to-catch-exception-caused-in-c-in-python, with the `try`-`catch` in `scipy/stats/_csparsetool.pyx`. That will keep the C++ sources free from Python C API code.

Reply:
On second thought: the line `#include <numpy/npy_common.h>` already pulls in `Python.h`, so if that's okay then this can simply be done here.